### ISSUE TITLE
Seq Lab: Update Pool ingest calc RNR/Remove threading/Write Rows so Smartsheet for Builder app

### ISSUE THREAD

**Issue by LMT337** [GTAC-MGI/GTAC-ESP-LIMS#486]
- [x] If value is in ingest doc, add value to Enitity : Requested Number of Reads (M), if not, do pool math
- [x] Remove threading for Ingest pipelines, race conditions causing issues
- [x] Write rows to Bulider sheet and Billing sheet in Smartsheet
- [x]  Integreate Builder functions into Pool/Library ingest Pipelines
- [x]  Requested reads testing
- [x]  Builder sheet updates from SEQ feedback

**Comment by LMT337** [comment]
Also check to make sure values are in each row if they are present for a pool and check that the sum equals the pool target.

**Comment by LMT337** [comment]
Targeted read-pairs (millions) for Library

**Comment by LMT337** [comment]
Just add the column Targeted read-pairs (millions) for Library (already in the library ingest sheet) to the pool ingest sheet.
Assign both the pool target (only needed in the first row if you choose) and the individual library target (needed for each library in the pool).
If the library target column does not have values, it will divide the total pool value evenly among the libraries.
If there are library target values, but one or more is missing ingest will fail.
If there are library target values, but they do not add up to the pool target value, ingest will fail.

**Comment by LMT337** [comment]
Threading removed, requested number of reads functions completed, neeed to test. 

Waiting on scheduling smarthsheets to begin development for writing rows from ingest.

**Comment by LMT337** [comment]
SR fields to send to builder development [sheet](https://app.smartsheet.com/sheets/4R8vfcvCrfF9v284JPwCcphhhGqQ5m4cFwqqXRJ1?view=grid)
[Scheduling_Fields.md](https://github.com/user-attachments/files/17847071/Scheduling_Fields.md)

**Comment by LMT337** [comment]
Billing [sheet](https://app.smartsheet.com/sheets/fwRFW8Qmj7Cwx4r45F5hGpcX9fVwWCcgx7hqHQp1?view=grid), follow up if this is the format moving forward, fields to write to sheet.

**Comment by LMT337** [comment]
Notes from Bill: 

- The inconsistency is mostly coming from mapping the builder columns from their box spreadsheets. My opinion is that the move should be toward conformity with ESP naming and away from box but then give seq lab final right of refusal over naming.
 
- Sample Name should be added to Builder and sourced from entity.  I know pools don’t have Sample Name, so left blank?
I think Pool Name should also map to the Illumina Library variable, for the libraries brought in as pools.
They use the Pool Group to track what they name it in the pooling protocol.What I called Pool ESP ID in the builder sheet is what I was using a placeholder for tracking pools that get made in the pooling workflow.  It could also be thought of / called New Pool Name
- 
- Sample Name = Illumina Library entity field
   Pool Name -> Customer Pool Name, for libraries brought in receive library pools
   Pool ESP ID -> GTAC Pool Name

- subdividing ESP ID clarifies the thing.
   Libraries brought in as Libraries will have a blank ESP Pool ID
   Pools ingested will have a value for ESP Pool ID and its libraries will have a ESP LIbrary ID
   Pools made in Seq Lab will get a value for ESP Pool ID manually entered or entered by pooling workflow

- sample data structure is still:
Pool ingest:
sr - parent
    pool- child
        library -child
        library -child
Library Ingest:
sr - parent
    library -child
    library -child

**Comment by LMT337** [comment]
Complete Worfklows (Receive Library/Pools) with pipeline script, waiting for direction on Reveiw Workflows without Pipeline scripts.

**Comment by LMT337** [comment]
Library Ingest/Pool ingest Scheduler sheet fuctions built, working. Waiting on feedback from SEQ on Column order, SR field values.

**Comment by LMT337** [comment]
Added Billing row function.  Waiting on feedback from SEQ on fields and column order.

**Comment by LMT337** [comment]
Add Average Library Size (bp) and Client Assessed Molarity (nM) to Builder sheet for Matt.

**Comment by LMT337** [comment]
Closing this ticket, if additional updates are required, make a new ticket.
